### PR TITLE
airbyte ci test to support --extras

### PR DESCRIPTION
**Pull request by erohmensing**
Our `airbyte-ci test` command supports [groups](https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups) but not [extras](https://python-poetry.org/docs/pyproject/#extras). This PR allows configuring extras to install if necessary (like when running the unit tests in the [upstack PR](https://github.com/airbytehq/airbyte/actions/runs/8446603559/job/23135651978?pr=36497)), and renames to better distinguish between the groups. 

We probably don't need to include `dev` as an optional group since it's implicitly installed, but I didn't mess with that here, for one because I wanted to test the rename.

For usage see [upstack PR.](https://github.com/airbytehq/airbyte/pull/36497/files#diff-1734d8b375bd2fc6fd8419f94ca88aad97eefaf2e78fa548c7379b28fca6456dR97-R101)

### CODE REVIEW

**Comment by erohmensing** [comment]
* **#36497** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36497?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#36527** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36527?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @erohmensing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: .github/workflows/airbyte-ci-tests.yml [hunk 1]
```diff
@@ -30,6 +30,7 @@ jobs:
         with:
           # Note: expressions within a filter are OR'ed
           filters: |
+            # This list is duplicated in `pipelines/airbyte_ci/test/__init__.py`
```
Comment (erohmensing) [inline .github/workflows/airbyte-ci-tests.yml:33]: This was just a note for me when I was working on the [upstack PR](https://github.com/airbytehq/airbyte/actions/runs/8446603559/job/23135651978?pr=36497)

There is some weird stuff around keepign this list up to date and also the output of changes being calculated differently from the `--modified` flag that we should look into later but I haven't addressed in this stack.

**Review by natikgadzhi — APPROVED** [review]

**Review by alafanechere — APPROVED** [review]
LGTM.
Nit stacking remark: I would have first renamed `extra_poetry_groups` to `optional_poetry_groups` in a PR and would have introduced the `poetry_extras` in a different one.
* What's exactly the difference between groups and extras? 
* Is there a way to install everything in a single poetry command instead of having packages to specify the same things?

**Comment by erohmensing** [comment]
@alafanechere 

> What's exactly the difference between groups and extras?

They are admittedly confusing. [Groups](https://python-poetry.org/docs/managing-dependencies/#dependency-groups) are only used to organize dependencies, and the very imporant thing is `Dependency groups, other than the implicit main group, must only contain dependencies you need in your development process. Installing them is only possible by using Poetry.`, i.e. they can only be dev dependencies. I _think_ they're basically like dev-extras. `To declare a set of dependencies, which add additional functionality to the project during runtime, use [extras](https://python-poetry.org/docs/pyproject/#extras) instead. Extras can be installed by the end user using pip.`, i.e. they are similar to setup.py extras in that they can contain code that is shipped with the package and can be installed by specifying extras. 

> Is there a way to install everything in a single poetry command instead of having packages to specify the same things?

Not that I can see. [There is no command to install all groups](https://github.com/orgs/python-poetry/discussions/7605), and installing groups is fundamentally different from installing extras. 

I guess I don't know if we really need to support groups at all, since we only use `dev` groups, and that one is implicitly always downloaded (e.g. `poetry install` == `poetry install --with dev`.

**Comment by alafanechere** [comment]
> I guess I don't know if we really need to support groups at all, since we only use dev groups, and that one is implicitly always downloaded (e.g. poetry install == poetry install --with dev.

You are right. I think I originally eagerly implemented this because I did not know that `poetry install` installed `dev` too and originally packages had different extras name in setup.py for dev deps: `tests` /  `dev` etc.

Thanks for the clarification on `group` vs `extras`.  I think it makes sense from a DX standpoint. You don't have to care too much when you run `poetry install` manually. And when you install packages at build time you can use a flag to ignore dev deps.

**Comment by erohmensing** [comment]
### Merge activity

* **Mar 27, 2:16 PM EDT**: @erohmensing started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/36527).

**Pull request merged**